### PR TITLE
Added RFC 6570 complaint form style query expansion as optional param… 

### PR DESCRIPTION
**Pull request by codecracker007**
# Add RFC 6570 support for optional parameters as form-style query expansions

## Motivation and Context

Adds optional parameter support with RFC 6570 compliance as form-style query expansions. This change addresses issue #378, .

## Breaking Changes
None

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

### CODE REVIEW

**Review by ihrpr — CHANGES_REQUESTED** [review]
Thank you for working on this!

This PR makes good progress toward [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570) support by adding form-style query expansion. 

There are a few suggestions: 

 1. Type Conversion for Query Parameters. Query parameters from URLs are always strings, but the code doesn't convert them to match function parameter
  types.

  Example of the problem:
  ```
  @server.resource("resource://items/{category}{?limit}")
  def get_items(category: str, limit: int = 10) -> str:
      # When called with resource://items/books?limit=20
      # limit will be string "20" not int 20
      return f"Items in {category}, limit: {limit} (type: {type(limit)})"
```
  Suggested fix: Add type conversion based on function annotations in 
  
 
  2. Handle Edge Cases in Query Parsing. The current implementation might not handle edge cases properly:
  - Empty query values (?param=)
  - URL-encoded values (?name=John%20Doe)

  3. Add Tests for Type Conversion Scenarios

  4. Consider Documenting Query Parameter Behavior

**Comment by codecracker007** [comment]
@ihrpr thanks for the feedback

during the fn call with params              [result = self.fn(**params)](https://github.com/modelcontextprotocol/python-sdk/blob/f2f4dbdcbd30fd00ced777fd1f59d00624362c97/src/mcp/server/fastmcp/resources/templates.py#L73)

can i use [call_fn_with_arg_validation](url) https://github.com/modelcontextprotocol/python-sdk/blob/f2f4dbdcbd30fd00ced777fd1f59d00624362c97/src/mcp/server/fastmcp/utilities/func_metadata.py#L45 for type validation and conversion and should i fallback to default values for optional parameters if validation fails for any one of them

**Comment by codecracker007** [comment]
@ihrpr the pydantic validate_call already takes care of the compatible type conversion based on annontated parameters for query parameters and raises  validation error for incompatible types added a decorator which check if the validation error is raised by optional parameters and ignores them so they fall back to default values to keep it compliant

added relevant tests covering these cases

for the cases in query parsing the empty values and url encoded parameters are handled by parse_qs added relevant tests covering these

**Review by felixweinberger — CHANGES_REQUESTED** [review]
Apologies for the delay in coming back to this - there seem to be some failing tests & merge conflicts, would you be able to resolve?

I'll prioritize responding quickly if you have the bandwidth to update the PR here.

File: src/mcp/server/fastmcp/utilities/func_metadata.py [hunk 1]
```diff
@@ -521,3 +523,110 @@ def _convert_to_content(
         result = pydantic_core.to_json(result, fallback=str, indent=2).decode()
 
     return [TextContent(type="text", text=result)]
+
+
+def use_defaults_on_optional_validation_error(
```
Comment (felixweinberger) [inline src/mcp/server/fastmcp/utilities/func_metadata.py:528]: We should fail fast if an optional parameter was provided with a non-matching type.

I think we should completely remove this as well as the call to it in `templates.py`

File: src/mcp/server/fastmcp/server.py [hunk 2]
```diff
@@ -514,15 +535,6 @@ def decorator(fn: AnyFunction) -> AnyFunction:
             has_func_params = bool(inspect.signature(fn).parameters)
 
             if has_uri_params or has_func_params:
-                # Validate that URI params match function params
-                uri_params = set(re.findall(r"{(\w+)}", uri))
-                func_params = set(inspect.signature(fn).parameters.keys())
-
-                if uri_params != func_params:
-                    raise ValueError(
-                        f"Mismatch between URI parameters {uri_params} and function parameters {func_params}"
```
Comment (felixweinberger) [inline src/mcp/server/fastmcp/server.py:523]: note: removing this validation makes sense because we now support optional params.

File: src/mcp/server/fastmcp/resources/templates.py [hunk 3]
```diff
@@ -34,40 +46,127 @@ def from_function(
...
+        # Get schema from TypeAdapter using the original function for correct schema
+        parameters = TypeAdapter(original_fn).json_schema()
 
-        # ensure the arguments are properly cast
-        fn = validate_call(fn)
+        # First, apply pydantic's validation and coercion
+        validated_fn = validate_call(original_fn)
+
+        # Then, apply our decorator to handle default fallback for optional params
+        final_fn = use_defaults_on_optional_validation_error(validated_fn)
```
Comment (felixweinberger) [inline src/mcp/server/fastmcp/resources/templates.py:61]: remove, we should fail fast and not silently when invalid params are provided by clients.

File: src/mcp/server/fastmcp/resources/templates.py [hunk 4]
```diff
@@ -34,40 +46,127 @@ def from_function(
...
+            # First add extracted parameters
+            fn_params = {
+                name: value
+                for name, value in params.items()
+                if name in self.required_params or name in self.optional_params
+            }
+
+            # self.fn is now multiply-decorated:
+            # 1. validate_call for coercion/validation
+            # 2. our new decorator for default fallback on optional param validation err
```
Comment (felixweinberger) [inline src/mcp/server/fastmcp/resources/templates.py:168]: don't use a fallback, just fail if the optional params are supplied incorrectly

File: tests/issues/test_141_resource_templates.py [hunk 5]
```diff
@@ -112,3 +146,113 @@ def get_user_profile(user_id: str) -> str:
 
         with pytest.raises(Exception):  # Specific exception type may vary
             await session.read_resource(AnyUrl("resource://users/123/invalid"))  # Invalid template
+
+
```
Comment (felixweinberger) [inline tests/issues/test_141_resource_templates.py:150]: thank you for adding significant test coverage - we'll want to update these to catch expected validation errors rather than the fallback logic.

File: src/mcp/server/fastmcp/resources/templates.py [hunk 6]
```diff
@@ -34,40 +46,127 @@ def from_function(
         mime_type: str | None = None,
     ) -> ResourceTemplate:
         """Create a template from a function."""
-        func_name = name or fn.__name__
+        original_fn = fn
```
Comment (felixweinberger) [inline src/mcp/server/fastmcp/resources/templates.py:49]: `original_fn`, `fn`, `final_fn` - this gets quite confusing.

I think if we remove the `use_defaults_on_optional_validation_error` we shouldn't need any of these and can just stick to `fn`

**Comment by Kludex** [comment]
Don't we want something more FastAPI-ish? Like `Annotated[type[T], Query()]`? 🤔

**Comment by felixweinberger** [comment]
> Don't we want something more FastAPI-ish? Like `Annotated[type[T], Query()]`? 🤔

Something like this?

```python
from typing import Annotated

@mcp.resource("search://{category}/{id}")
def search(
    category: Annotated[str, Path()],
    id: Annotated[int, Path()],
    format: Annotated[str, Query()] = "json",
    limit: Annotated[int, Query()] = 10
):
    ...
```

Note: there's also https://github.com/modelcontextprotocol/python-sdk/pull/1439 which is touching the same area so we'd need some coordination cc: @beaterblank

**Comment by Kludex** [comment]
> > Don't we want something more FastAPI-ish? Like `Annotated[type[T], Query()]`? 🤔
> 
> Something like this?
> 
> ```python
> from typing import Annotated
> 
> @mcp.resource("search://{category}/{id}")
> def search(
>     category: Annotated[str, Path()],
>     id: Annotated[int, Path()],
>     format: Annotated[str, Query()] = "json",
>     limit: Annotated[int, Query()] = 10
> ):
>     ...
> ```
> 
> Note: there's also #1439 which is touching the same area so we'd need some coordination cc: @beaterblank

For `category` and `id`, you wouldn't need it, but you could... Also, I think `Query`/`Path` would have the `default` parameter, so you don't need to add default in the parameters.

**Comment by felixweinberger** [comment]
> > > Don't we want something more FastAPI-ish? Like `Annotated[type[T], Query()]`? 🤔
> > 
> > 
> > Something like this?
> > ```python
> > from typing import Annotated
> > 
> > @mcp.resource("search://{category}/{id}")
> > def search(
> >     category: Annotated[str, Path()],
> >     id: Annotated[int, Path()],
> >     format: Annotated[str, Query()] = "json",
> >     limit: Annotated[int, Query()] = 10
> > ):
> >     ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Note: there's also #1439 which is touching the same area so we'd need some coordination cc: @beaterblank
> 
> For `category` and `id`, you wouldn't need it, but you could... Also, I think `Query`/`Path` would have the `default` parameter, so you don't need to add default in the parameters.

OK thanks, so more like this?

```python
from typing import Annotated

@mcp.resource("search://{category}/{id}")
def search(
    category: str,
    id: int,
    format: Annotated[str, Query(default="json")],
    limit: Annotated[int, Query(default=10)],
):
    ...
```

**Comment by beaterblank** [comment]
> > Don't we want something more FastAPI-ish? Like `Annotated[type[T], Query()]`? 🤔
> 
> Something like this?
> 
> ```python
> from typing import Annotated
> 
> @mcp.resource("search://{category}/{id}")
> def search(
>     category: Annotated[str, Path()],
>     id: Annotated[int, Path()],
>     format: Annotated[str, Query()] = "json",
>     limit: Annotated[int, Query()] = 10
> ):
>     ...
> ```
> 
> Note: there's also #1439, which is touching the same area, so we'd need some coordination cc: @beaterblank

Yes, using `Annotated` sounds like a good approach, but we should ideally support both styles. 

~Also, a small nitpick, the classes should be named `PathParameter` and `QueryParameter`, since I spent a few minutes confused :)~ or maybe its best stick with how fast api does it.

I also noticed an issue in my PR: I’m currently inferring types only from the URI, but this should be done based on the function signature value's types instead. can reuse the code from this PR for that.

Additionally, since both features modify the same function in different ways, we should refactor the logic into smaller, more focused functions. That will make the codebase cleaner and easier to extend in the future.

At this point, I could either merge this PR into mine and resolve the conflicts properly, or do it the other way around.



merging this into #1439

**Comment by beaterblank** [comment]
Can close this PR merged into #1439 .

**Comment by felixweinberger** [comment]
Consolidating discussion on https://github.com/modelcontextprotocol/python-sdk/pull/1439
